File: dasboard/src/utils/dateUtils.js
/**
 * Date and Time Utilities for India/Kerala (IST - UTC+5:30)
 */

// IST timezone constant
const IST_TIMEZONE = 'Asia/Kolkata';

/**
 * Format date in IST timezone
 * @param {string|Date} dateString - Date string or Date object
 * @param {object} options - Intl.DateTimeFormat options
 * @returns {string} Formatted date string
 */
export const formatDateIST = (dateString, options = {}) => {
  if (!dateString) return '-';
  
  const date = typeof dateString === 'string' ? new Date(dateString) : dateString;
  
  if (isNaN(date.getTime())) return '-';
  
  const defaultOptions = {
    timeZone: IST_TIMEZONE,
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    ...options
  };
  
  return new Intl.DateTimeFormat('en-IN', defaultOptions).format(date);
};

/**
 * Format date and time in IST timezone
 * @param {string|Date} dateString - Date string or Date object
 * @param {object} options - Intl.DateTimeFormat options
 * @returns {string} Formatted date and time string
 */
export const formatDateTimeIST = (dateString, options = {}) => {
  if (!dateString) return '-';
  
  const date = typeof dateString === 'string' ? new Date(dateString) : dateString;
  
  if (isNaN(date.getTime())) return '-';
  
  const defaultOptions = {
    timeZone: IST_TIMEZONE,
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: true,
    ...options
  };
  
  return new Intl.DateTimeFormat('en-IN', defaultOptions).format(date);
};

/**
 * Format time only in IST timezone
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted time string
 */
export const formatTimeIST = (dateString) => {
  if (!dateString) return '-';
  
  const date = typeof dateString === 'string' ? new Date(dateString) : dateString;
  
  if (isNaN(date.getTime())) return '-';
  
  return new Intl.DateTimeFormat('en-IN', {
    timeZone: IST_TIMEZONE,
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: true
  }).format(date);
};

/**
 * Get current date in IST timezone (YYYY-MM-DD format for date inputs)
 * @returns {string} Current date in YYYY-MM-DD format
 */
export const getCurrentDateIST = () => {
  const now = new Date();
  const istDate = new Date(now.toLocaleString('en-US', { timeZone: IST_TIMEZONE }));
  
  const year = istDate.getFullYear();
  const month = String(istDate.getMonth() + 1).padStart(2, '0');
  const day = String(istDate.getDate()).padStart(2, '0');
  
  return `${year}-${month}-${day}`;
};

/**
 * Get current datetime in IST timezone (ISO format)
 * @returns {string} Current datetime in ISO format
 */
export const getCurrentDateTimeIST = () => {
  const now = new Date();
  const istDate = new Date(now.toLocaleString('en-US', { timeZone: IST_TIMEZONE }));
  return istDate.toISOString();
};

/**
 * Convert date to IST and return as ISO string
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} ISO string in IST
 */
export const toISTISO = (dateString) => {
  if (!dateString) return null;
  
  const date = typeof dateString === 'string' ? new Date(dateString) : dateString;
  
  if (isNaN(date.getTime())) return null;
  
  // Get IST time
  const istString = date.toLocaleString('en-US', { timeZone: IST_TIMEZONE });
  const istDate = new Date(istString);
  
  return istDate.toISOString();
};

/**
 * Format date for display (short format: DD MMM YYYY)
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted date
 */
export const formatDateShort = (dateString) => {
  return formatDateIST(dateString, {
    day: '2-digit',
    month: 'short',
    year: 'numeric'
  });
};

/**
 * Format date for display (long format: DD MMMM YYYY)
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted date
 */
export const formatDateLong = (dateString) => {
  return formatDateIST(dateString, {
    day: '2-digit',
    month: 'long',
    year: 'numeric'
  });
};

/**
 * Format datetime for display (DD MMM YYYY, HH:MM AM/PM)
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted datetime
 */
export const formatDateTimeShort = (dateString) => {
  return formatDateTimeIST(dateString, {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: true
  });
};

/**
 * Format datetime for display (DD MMMM YYYY, HH:MM:SS AM/PM)
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted datetime
 */
export const formatDateTimeLong = (dateString) => {
  return formatDateTimeIST(dateString, {
    day: '2-digit',
    month: 'long',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: true
  });
};

/**
 * Get relative time (e.g., "2 hours ago", "yesterday")
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Relative time string
 */
export const getRelativeTime = (dateString) => {
  if (!dateString) return '-';
  
  const date = typeof dateString === 'string' ? new Date(dateString) : dateString;
  
  if (isNaN(date.getTime())) return '-';
  
  const now = new Date();
  const nowIST = new Date(now.toLocaleString('en-US', { timeZone: IST_TIMEZONE }));
  const dateIST = new Date(date.toLocaleString('en-US', { timeZone: IST_TIMEZONE }));
  
  const diffMs = nowIST - dateIST;
  const diffSecs = Math.floor(diffMs / 1000);
  const diffMins = Math.floor(diffSecs / 60);
  const diffHours = Math.floor(diffMins / 60);
  const diffDays = Math.floor(diffHours / 24);
  
  if (diffSecs < 60) return 'Just now';
  if (diffMins < 60) return `${diffMins} minute${diffMins > 1 ? 's' : ''} ago`;
  if (diffHours < 24) return `${diffHours} hour${diffHours > 1 ? 's' : ''} ago`;
  if (diffDays === 1) return 'Yesterday';
  if (diffDays < 7) return `${diffDays} days ago`;
  
  return formatDateShort(dateString);
};

export default {
  formatDateIST,
  formatDateTimeIST,
  formatTimeIST,
  getCurrentDateIST,
  getCurrentDateTimeIST,
  toISTISO,
  formatDateShort,
  formatDateLong,
  formatDateTimeShort,
  formatDateTimeLong,
  getRelativeTime,
  IST_TIMEZONE
};

File: ResortApp/app/utils/date_utils.py
"""
Date and Time Utilities for India/Kerala (IST - UTC+5:30)
"""
from datetime import datetime, timezone, timedelta
from typing import Optional, Union
import pytz

# IST timezone constant
IST_TIMEZONE = pytz.timezone('Asia/Kolkata')

def get_ist_now() -> datetime:
    """
    Get current datetime in IST timezone
    
    Returns:
        datetime: Current datetime in IST
    """
    return datetime.now(IST_TIMEZONE)

def get_ist_today() -> datetime:
    """
    Get today's date at midnight in IST timezone
    
    Returns:
        datetime: Today at 00:00:00 IST
    """
    now_ist = get_ist_now()
    return now_ist.replace(hour=0, minute=0, second=0, microsecond=0)

def to_ist(dt: Union[datetime, str, None]) -> Optional[datetime]:
    """
    Convert a datetime to IST timezone
    
    Args:
        dt: datetime object, ISO string, or None
        
    Returns:
        datetime: Datetime in IST timezone, or None if input is None
    """
    if dt is None:
        return None
    
    # If string, parse it
    if isinstance(dt, str):
        try:
            # Try parsing ISO format
            dt = datetime.fromisoformat(dt.replace('Z', '+00:00'))
        except ValueError:
            # Try other common formats
            try:
                dt = datetime.strptime(dt, '%Y-%m-%d %H:%M:%S')
            except ValueError:
                dt = datetime.strptime(dt, '%Y-%m-%d')
    
    # If datetime is naive (no timezone), assume it's already in IST
    if dt.tzinfo is None:
        return IST_TIMEZONE.localize(dt)
    
    # Convert to IST
    return dt.astimezone(IST_TIMEZONE)

def format_datetime_ist(dt: Union[datetime, str, None], format_str: str = '%d %b %Y, %I:%M %p') -> str:
    """
    Format datetime in IST timezone
    
    Args:
        dt: datetime object, ISO string, or None
        format_str: Format string (default: 'DD MMM YYYY, HH:MM AM/PM')
        
    Returns:
        str: Formatted datetime string, or '-' if None
    """
    if dt is None:
        return '-'
    
    dt_ist = to_ist(dt)
    if dt_ist is None:
        return '-'
    
    return dt_ist.strftime(format_str)

def format_date_ist(dt: Union[datetime, str, None], format_str: str = '%d %b %Y') -> str:
    """
    Format date in IST timezone
    
    Args:
        dt: datetime object, ISO string, or None
        format_str: Format string (default: 'DD MMM YYYY')
        
    Returns:
        str: Formatted date string, or '-' if None
    """
    if dt is None:
        return '-'
    
    dt_ist = to_ist(dt)
    if dt_ist is None:
        return '-'
    
    return dt_ist.strftime(format_str)

def format_time_ist(dt: Union[datetime, str, None], format_str: str = '%I:%M %p') -> str:
    """
    Format time in IST timezone
    
    Args:
        dt: datetime object, ISO string, or None
        format_str: Format string (default: 'HH:MM AM/PM')
        
    Returns:
        str: Formatted time string, or '-' if None
    """
    if dt is None:
        return '-'
    
    dt_ist = to_ist(dt)
    if dt_ist is None:
        return '-'
    
    return dt_ist.strftime(format_str)

def get_ist_date_range(period: str = 'today'):
    """
    Get date range in IST timezone
    
    Args:
        period: 'today', 'week', 'month', 'year'
        
    Returns:
        tuple: (start_date, end_date) as datetime objects in IST
    """
    today_ist = get_ist_today()
    
    if period == 'today':
        start_date = today_ist
        end_date = today_ist + timedelta(days=1)
    elif period == 'week':
        # Start of week (Monday)
        days_since_monday = today_ist.weekday()
        start_date = today_ist - timedelta(days=days_since_monday)
        end_date = start_date + timedelta(days=7)
    elif period == 'month':
        # Start of month
        start_date = today_ist.replace(day=1)
        # Start of next month
        if start_date.month == 12:
            end_date = start_date.replace(year=start_date.year + 1, month=1)
        else:
            end_date = start_date.replace(month=start_date.month + 1)
    elif period == 'year':
        # Start of year
        start_date = today_ist.replace(month=1, day=1)
        # Start of next year
        end_date = start_date.replace(year=start_date.year + 1)
    else:
        start_date = today_ist
        end_date = today_ist + timedelta(days=1)
    
    return start_date, end_date

def utc_to_ist(utc_dt: datetime) -> datetime:
    """
    Convert UTC datetime to IST
    
    Args:
        utc_dt: UTC datetime object
        
    Returns:
        datetime: Datetime in IST
    """
    if utc_dt.tzinfo is None:
        # Assume UTC if naive
        utc_dt = utc_dt.replace(tzinfo=timezone.utc)
    
    return utc_dt.astimezone(IST_TIMEZONE)

def ist_to_utc(ist_dt: datetime) -> datetime:
    """
    Convert IST datetime to UTC
    
    Args:
        ist_dt: IST datetime object
        
    Returns:
        datetime: Datetime in UTC
    """
    if ist_dt.tzinfo is None:
        # Assume IST if naive
        ist_dt = IST_TIMEZONE.localize(ist_dt)
    
    return ist_dt.astimezone(timezone.utc)



